File: FortyTwo/Shared/Models/Domino.cs
using Standart.Hash.xxHash;
using System;
using System.Text;
using System.Text.Json.Serialization;

namespace FortyTwo.Shared.Models
{
    public enum Orientation { Left, Right };

    public class Domino : IEquatable<Domino>
    {
        [JsonConstructor]
        public Domino(string id, int top, int bottom, Orientation orientation)
        {
            Id = id;
            Top = top;
            Bottom = bottom;
            Orientation = orientation;
        }

        public Domino(int top, int bottom, Orientation orientation = Orientation.Left)
        {
            if (top < 0) throw new ArgumentOutOfRangeException(nameof(top));
            if (bottom < 0) throw new ArgumentOutOfRangeException(nameof(bottom));

            if (top > bottom) throw new ArgumentOutOfRangeException("Invalid Domino creation.");

            Top = top;
            Bottom = bottom;

            var bytes = Encoding.UTF8.GetBytes($"{Top}/{Bottom}");
            Id = xxHash64.ComputeHash(bytes, bytes.Length).ToString();
        }

        public string Id { get; }
        public int Top { get; }
        public int Bottom { get; }
        public Orientation Orientation { get; set; }
        public int Value => ((Top + Bottom) % 5 == 0) ? (Top + Bottom) : 0;

        public bool IsDouble => this.Top == this.Bottom;

        public Suit GetSuit(Suit trump)
            => IsOfSuit(trump)
                ? trump
                : (Suit)(Math.Max(Top, Bottom));

        public bool IsOfSuit(Suit suit, Suit? trump = null)
            => trump == null || suit == trump
                ? (this.Top == (int)suit || this.Bottom == (int)suit)
                : (this.Top == (int)suit && this.Bottom != (int)trump ) || (this.Bottom == (int)suit && this.Top != (int)trump);

        public int GetSuitValue(Suit suit, Suit trump)
            => IsOfSuit(suit, trump)
                ? IsDouble ? 7 : this.Top == (int)suit ? this.Bottom : this.Top
                : IsOfSuit(trump)
                    ? 10 + (IsDouble ? 7 : this.Top == (int)suit ? this.Bottom : this.Top)
                    : -1;

        public bool Equals(Domino other)
        {
            return (this.Top == other.Top && this.Bottom == other.Bottom)
                || (this.Top == other.Bottom && this.Bottom == other.Top);
        }
    }
}

File: cloudflare/packages/rules/src/domino.ts
import { Suit } from './suit';

export interface Domino {
  readonly id: string;
  readonly top: number;
  readonly bottom: number;
}

export function createDomino(top: number, bottom: number): Domino {
  if (top < 0 || bottom < 0) throw new RangeError('Domino halves must be non-negative');
  const [lo, hi] = top <= bottom ? [top, bottom] : [bottom, top];
  return { id: `${lo}/${hi}`, top: lo, bottom: hi };
}

export function dominoValue(d: Domino): number {
  const sum = d.top + d.bottom;
  return sum % 5 === 0 ? sum : 0;
}

export function isDouble(d: Domino): boolean {
  return d.top === d.bottom;
}

export function isOfSuit(d: Domino, suit: Suit, trump?: Suit | null): boolean {
  if (trump == null || suit === trump) {
    return d.top === suit || d.bottom === suit;
  }
  return (d.top === suit && d.bottom !== trump) || (d.bottom === suit && d.top !== trump);
}

export function getSuit(d: Domino, trump: Suit): Suit {
  return isOfSuit(d, trump) ? trump : (Math.max(d.top, d.bottom) as Suit);
}

export function getSuitValue(d: Domino, suit: Suit, trump: Suit): number {
  if (isOfSuit(d, suit, trump)) {
    return isDouble(d) ? 7 : d.top === suit ? d.bottom : d.top;
  }
  if (isOfSuit(d, trump)) {
    return 10 + (isDouble(d) ? 7 : d.top === suit ? d.bottom : d.top);
  }
  return -1;
}

export function dominoEquals(a: Domino, b: Domino): boolean {
  return (a.top === b.top && a.bottom === b.bottom) || (a.top === b.bottom && a.bottom === b.top);
}
